src/KafkaFlow/Consumers/KafkaConsumerFlowManager.cs

```diff
@@ -0,0 +1,109 @@
+namespace KafkaFlow.Consumers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Confluent.Kafka;
+
+    internal class KafkaConsumerFlowManager : IKafkaConsumerFlowManager
+    {
+        private readonly IConsumer<byte[], byte[]> consumer;
+        private readonly CancellationToken stopCancellationToken;
+        private readonly ILogHandler logHandler;
+
+        private CancellationTokenSource heartbeatTokenSource;
+        private Task heartbeatTask;
+
+        private readonly List<TopicPartition> paused = new List<TopicPartition>();
+
+        public KafkaConsumerFlowManager(
+            IConsumer<byte[], byte[]> consumer,
+            CancellationToken stopCancellationToken,
+            ILogHandler logHandler)
+        {
+            this.consumer = consumer;
+            this.stopCancellationToken = stopCancellationToken;
+            this.logHandler = logHandler;
+        }
+
+        public void Pause(IReadOnlyCollection<TopicPartition> topicPartitions)
+        {
+            lock (this.paused)
+            {
+                topicPartitions = topicPartitions.Except(this.paused).ToList();
+
+                if (!topicPartitions.Any())
+                {
+                    return;
+                }
+
+                this.consumer.Pause(topicPartitions);
+                this.paused.AddRange(topicPartitions);
+
+                if (!this.AreAllPartitionsPaused())
```

Comment:
I think would be more readable if we change to `HasUnpausedPartition` and return in case of true.

Reply:
changed to `HasRunningPartitions()`